documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,888 @@
+
+For new connections, after all packet transformations are applied, we will create a new flow in the flow table.
+
+Since a connection can be bi-directional, the flows needs to be created as flow pairs: the one in the same direction as the packet is usually called forwarding flow, while the other direction is called reverse flow. Of course, the flow pairs **MUST** be stored based on the pipeline direction.
+
+- When outbound pipeline creates a flow, the forwarding flow should be created on the outbound side, while reverse flow should be created on the inbound side.
+- When inbound pipeline creates a flow, it creates the flow in reversed way.
+
+And here is how to create the flow pairs:
+
+- The forwarding flow creation is straight forward, because all the final transformations are defined in the flow actions and metadata bus.
+- To properly create the reverse flow, we will learn the information from the original tunnels, and reverse them as return encaps, which is why we need to save all the original tunnel information when doing packet decaps.
+
+##### 5.6.2.1. Tunnel learning
+
+You might already notice that, although we uses the outer encap from the original packet to create the reverse flow, but the outer encaps are decap'ed and not used in flow matching. This creates a problem when source side fails over. Like the graph shows below, the traffic shifted from VTEP 1 (Green side) to VTEP 2 (Blue side), which changes the source IP in the encap. If we don't do anything, the existing reverse flow will continue to send the packet back to green side and causing the traffic being dropped.
+
+![Tunnel learning](./images/dash-tunnel-learning.svg)
+
+To address this problem, the source information of each encap **MUST** be saved in the flow and compared during flow match. These information should no be part of the key, but whenever they change, we should reprocess the packet and update the saved reverse tunnel in the existing flow.
+
+##### 5.6.2.2. Asymmetrical encap handling
+
+There could be cases where the encaps are asymmetrical, which means the incoming packet and return packet uses different encaps. For example, in the case below, the encap of step 2 and 4 are not symmetrical. This means we could never recreate the flow on the DASH pipeline 2 side to tunnel the packet into the extra hop first with our current flow creation logic.
+
+![Asymmetrical encap](./images/dash-asymmetrical-encap.svg)
+
+To fix this issue, a special action called `reverse_tunnel` is defined, which enables reverse side of the encaps.
+
+#### 5.6.3. Flow resimulation
+
+When certain policies are changed, we might need to update the flows that is associated with the policies. For example, when a VM is moved to another place, the VNET mapping entry will be changed, and we need to update the existing flows that is associated with the VNET mapping entry. Otherwise, the outgoing traffic will be tunneled to incorrect place. This is called flow resimulation.
```

Comment:
```suggestion
In the event of a policy change, we might need to update the flows associated with the policies. For example, when a VM is moved to another location, the VNET mapping entry will be changed, and we need to update the existing flows associated with the VNET mapping entry. Otherwise, the outgoing traffic will be tunneled to an incorrect location. This is called flow resimulation.
```